src/primitives/encode.rs

```diff
@@ -0,0 +1,311 @@
+// SPDX-License-Identifier: MIT
+
+//! Bech32 address encoding.
+//!
+//! This module provides types and iterators that can be used to encode data as a bech32 address in
+//! a variety of ways without any allocations, generating, verifying, and appending checksums,
+//! prepending HRP strings etc.
+//!
+//! In general, directly using these adaptors is not very ergonomic, and users are recommended to
+//! instead use the higher-level functions at the root of this crate.
+//!
+//! # Examples
+//!
+//! ```
+//! use bech32::{Bech32, ByteIterExt, Fe32IterExt, Fe32, Hrp};
+//!
+//! let witness_prog = [
+//!     0x75, 0x1e, 0x76, 0xe8, 0x19, 0x91, 0x96, 0xd4,
+//!     0x54, 0x94, 0x1c, 0x45, 0xd1, 0xb3, 0xa3, 0x23,
+//!     0xf1, 0x43, 0x3b, 0xd6,
+//! ];
+//!
+//! // Get a stream of characters representing the bech32 encoded address
+//! // use "bc" for the human-readable part.
+//! let hrp = Hrp::parse("bc").expect("bc is valid hrp string");
+//! let chars = witness_prog
+//!     .iter()
+//!     .copied()
+//!     .bytes_to_fes()
+//!     .encoder::<Bech32>(&hrp)
+//!     .prepend_witness_version(Fe32::Q) // Optionally add witness version.
+//!     .chars();
+//!
+//! #[cfg(feature = "alloc")]
+//! {
+//!     let addr = chars.collect::<String>();
+//!     assert_eq!(addr.to_uppercase(), "BC1QW508D6QEJXTDG4Y5R3ZARVARY0C5XW7KV8F3T4");
+//! }
+//! ```
+
+use core::iter::Iterator;
+use core::marker::PhantomData;
+
+use crate::primitives::checksum::HrpChecksumFe32Iter;
+use crate::primitives::hrp::{self, Hrp};
+use crate::primitives::iter::Checksummed;
+use crate::{Checksum, Fe32};
+
+/// The `Encoder` builds iterators that can be used to encode field elements into a bech32 address.
+#[derive(Clone, PartialEq, Eq)]
+pub struct Encoder<'hrp, I, Ck>
+where
+    I: Iterator<Item = Fe32>,
+    Ck: Checksum,
+{
+    /// The field elements to encode.
+    data: I,
+    /// The human-readable part used at the front of the address encoding.
+    hrp: &'hrp Hrp,
+    /// The witness version, if present.
+    witness_version: Option<Fe32>,
+    /// Checksum marker.
+    marker: PhantomData<Ck>,
+}
+
+impl<'hrp, I, Ck> Encoder<'hrp, I, Ck>
+where
+    I: Iterator<Item = Fe32>,
+    Ck: Checksum,
+{
+    /// Constructs a new bech32 encoder.
+    pub fn new(data: I, hrp: &'hrp Hrp) -> Self {
+        Self { data, hrp, witness_version: None, marker: PhantomData::<Ck> }
+    }
+
+    /// Add `witness_version` to the encoder (as first byte of encoded data).
+    ///
+    /// Note, caller to guarantee that witness version is within valid range (0-16).
+    pub fn prepend_witness_version(mut self, witness_version: Fe32) -> Self {
```

Comment:
In 5c9b549a3738cec82f633c14e274854b82edce45:

I'd panic here if `self.witness_version` is already set. Or just rename this method to `with_witness_version` or something.